src/parallel/parallel.cpp

```diff
@@ -123,32 +123,192 @@ ThreadManager::get_segment_index(size_t index)
+            auto [prev_key, prev_val] = this->compare_and_set_key_val(home, prev_key_copy, this->buckets_[idx].key, prev_val_copy);
+            if (prev_key == key) {
+                return;
+            }
+        }
+    }
+    auto [is_inserted, is_found] = this->distance_zero_insert(key, value, home);
+    if(is_inserted) {
+        return;
+    }
+    if(is_found) { //not sure about these?
+        return;
+    }
+    size_t offset = 0;
+    ParallelBucket entry_to_insert = {.key = key, .value = value, .hashcode = hashcode, .offset = offset}; //offset is arbitrary
+    auto [next_index, found] = this->find_next_index_lock(manager, home, key, offset);
+
+    if(found){
+        this->buckets_[next_index].value = value;
+        manager.release_all_locks();
+        return;
+    }
+
+    this->locked_insert(entry_to_insert, next_index);
+    manager.release_all_locks();
+    return;
 }
 
 bool
 ParallelRobinHoodHashTable::remove(KeyType key, ValueType value)
 {
-    // TODO
+    HashCodeType hashcode = this->hash(key);
+    size_t home = this->get_home(hashcode, this->capacity_);
+    ThreadManager manager = this->get_thread_lock_manager();
+    ParallelBucket entry_to_insert = {.key = key, .value = value, .hashcode = hashcode, .offset = 0}; //offset is arbitrary
+
+    auto[index, found] = this->find_next_index_lock(manager, home, key, entry_to_insert.offset);
+
+    if(!found) {
+        manager.release_all_locks();
+        return false;
+    }
+
+    size_t next_index = index++;
+    size_t curr_index = index;
+    
+    manager.lock(curr_index);
+    while(this->buckets_[next_index].offset > 0 && !entry_to_insert.is_empty()) {
+        manager.lock(next_index); 
+        ParallelBucket &entry_to_move = this->do_atomic_swap(entry_to_insert, next_index); //na this is fucked up 
+        this->buckets_[curr_index].key = entry_to_move.key;
+        this->buckets_[curr_index].value = entry_to_move.value;
+        this->buckets_[curr_index].hashcode = entry_to_move.hashcode;
+        this->buckets_[curr_index].offset = entry_to_move.offset--;
+        curr_index = next_index;
+        ++next_index;
+    }
+
+    entry_to_insert.unlock();
+    ParallelBucket empty_entry; //does this do the default vals? 
+    this->buckets_[curr_index].key = empty_entry.key;
+    this->buckets_[curr_index].value = empty_entry.value;
+    this->buckets_[curr_index].hashcode = empty_entry.hashcode;
+    this->buckets_[curr_index].offset = empty_entry.offset;
+    manager.release_all_locks();
+    return true;
 }
 
 std::pair<ValueType, bool>
 ParallelRobinHoodHashTable::find(KeyType key)
 {
-    // TODO
+    HashCodeType hashcode =this->hash(key);
+    size_t home = this->get_home(hashcode, this->capacity_);
+    ParallelBucket &zero_distance_key_pair = atomic_load_key_index(home);
+
+    if(zero_distance_key_pair.key == key) {
+        return {zero_distance_key_pair.key, true};
+    }
+    if(zero_distance_key_pair.offset == 0) {
```

Comment:
```suggestion
    if(zero_distance_key_pair.offset == SIZE_MAX) {
```
Offset 0 means home. I think you mean SIZE_MAX.